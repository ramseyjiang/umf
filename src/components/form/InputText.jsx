import React, {useState} from "react";
import PropTypes from 'prop-types'
import { Form } from 'react-bootstrap';

const InputText = ({
  name,
  type,
  placeholder,
  onChange,
  onBlur,
  min,
  max,
  error='',
  label,
  multiple,
  disabled,
  value
}) => {
  const [hidden, setHidden] = useState(false);

  const handleClick = e => {
    hidden === false ? setHidden(true) : setHidden(false);
  }

  return (
    <Form.Group>
      <Form.Label htmlFor={name}>{label}</Form.Label>
      <Form.Control className={error.length ? "is-invalid" : ""} type={hidden ? "text" : type} placeholder={placeholder} minLength={min} maxLength={max} name={name} value={value} onChange={onChange} disabled={disabled} required/>
      {name==="password" && <Form.Check type="checkbox" label={hidden ? "Hide password" : "Show password"} onClick={handleClick} />}
      {error.length && <div className="invalid-feedback">
        {error}
      </div>}
    </Form.Group>
  );
};

InputText.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string,
  placeholder: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
  min: PropTypes.number,
  max: PropTypes.number,
  error: PropTypes.string,
  label: PropTypes.string,
  multiple: PropTypes.bool,
  disabled: PropTypes.bool,
  value: PropTypes.string,
}

InputText.defaultProps = {
  name: 'name',
  type: 'text',
  placeholder: '',
  onChange: ()=>{},
  onBlur: ()=>{},
  min: 6,
  max: 100,
  error: '',
  label: '',
  multiple: false,
  disabled: false,
  value: ''
}

export default InputText;
