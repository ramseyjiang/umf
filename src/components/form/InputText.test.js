import React from "react";
import renderer from "react-test-renderer";
import InputText from "./InputText";
import { render, screen, cleanup } from "@testing-library/react";
import { mount } from "enzyme";

describe("Input Text component test", () => {
  afterEach(cleanup);

  it("render correctly input text component", () => {
    const InputTextComponent = renderer.create(<InputText />).toJSON();
    expect(InputTextComponent).toMatchSnapshot();
  });

  it("input fields correct", () => {
    const props = {
      name: "test name",
      label: "test label",
      placeholder: "test placeholder",
      value: "test input value",
    };

    const { getByText, getByPlaceholderText } = render(
      <InputText {...props} />
    );

    expect(getByText("test label")).toBeInTheDocument();
    expect(getByPlaceholderText("test placeholder")).toBeInTheDocument();
    expect(
      screen.getByDisplayValue("test input value", {
        exact: true,
      }).value
    ).toBe("test input value");
  });

  it("check the onChange callback", () => {
    const onChange = jest.fn();
    const props = {
      value: "input test",
      label: "First Name",
      name: "first_name",
      onChange,
    };
    const wrapper = mount(<InputText {...props} />);
    const label = wrapper.find("label");
    expect(label).toHaveLength(1);
    expect(label.text()).toEqual("First Name");

    const input = wrapper.find("input");
    expect(input).toHaveLength(1);
    expect(input.prop("type")).toEqual("text");
    expect(input.prop("name")).toEqual("first_name");

    input.simulate("change", { target: { value: "input test" } });
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it("renders email input with label given the type", () => {
    const props = {
      type: "email",
      label: "Email",
      name: "email",
    };

    const wrapper = mount(<InputText {...props} />);
    const label = wrapper.find("label");
    expect(label).toHaveLength(1);
    expect(label.prop("htmlFor")).toEqual("email");
    expect(label.text()).toEqual("Email");

    const input = wrapper.find("input");
    expect(input).toHaveLength(1);
    expect(input.prop("type")).toEqual("email");
    expect(input.prop("name")).toEqual("email");
  });

  it("test input error is correct", () => {
    const props = {
      type: "email",
      label: "Email",
      name: "email",
      error: "you input email is invalid",
    };

    const wrapper = mount(<InputText {...props} />);

    const invalidFb = wrapper.find(".invalid-feedback");
    expect(invalidFb.hasClass("invalid-feedback")).toBe(true);
    expect(invalidFb.text()).toEqual(props.error);
  });
});
