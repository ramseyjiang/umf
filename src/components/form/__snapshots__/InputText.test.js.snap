// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Text component test render correctly input text component 1`] = `
<div
  className="form-group"
>
  <label
    className="form-label"
    htmlFor="name"
  >
    
  </label>
  <input
    className="form-control"
    disabled={false}
    maxLength={100}
    minLength={6}
    name="name"
    onChange={[Function]}
    placeholder=""
    required={true}
    type="text"
    value=""
  />
  0
</div>
`;
